packages/messaging/src/controllers/sw-controller.ts

```diff
@@ -246,15 +246,25 @@ export class SwController implements FirebaseMessaging, FirebaseService {
     event.stopImmediatePropagation();
     event.notification.close();
 
+    // Note clicking on a notification with no link set will focus the Chrome's current tab.
     const link = getLink(internalPayload);
     if (!link) {
       return;
     }
 
-    let client = await getWindowClient(link);
+    // FM should only open/focus links from app's origin.
+    const url = new URL(link, self.location.href);
+    const originUrl = new URL(self.location.origin);
+
+    if (url.host !== originUrl.host) {
```

Comment:
Wouldn't it be a breaking change

Reply:
I guess it's not if you consider it as a security bug.

Reply:
Yay 🐛